server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -419,39 +428,27 @@ static final class AsyncAfterWriteAction {
             @Nullable final Translog.Location location,
             final RespondingWriteResult respond,
             final Logger logger,
+            @Nullable final PostWriteRefresh postWriteRefresh,
             @Nullable final Consumer<Runnable> postWriteAction
         ) {
             this.indexShard = indexShard;
             this.request = request;
-            boolean waitUntilRefresh = false;
-            switch (request.getRefreshPolicy()) {
-                case IMMEDIATE:
-                    indexShard.refresh("refresh_flag_index");
-                    refreshed.set(true);
-                    break;
-                case WAIT_UNTIL:
-                    if (location != null) {
-                        waitUntilRefresh = true;
-                        pendingOps.incrementAndGet();
-                    }
-                    break;
-                case NONE:
-                    break;
-                default:
-                    throw new IllegalArgumentException("unknown refresh policy: " + request.getRefreshPolicy());
-            }
-            this.waitUntilRefresh = waitUntilRefresh;
+            this.needsRefreshAction = request.getRefreshPolicy() != WriteRequest.RefreshPolicy.NONE;
             this.respond = respond;
             this.location = location;
+            this.postWriteRefresh = postWriteRefresh;
             this.postWriteAction = postWriteAction;
+            if (needsRefreshAction) {
+                pendingOps.incrementAndGet();
```

Comment:
Is there a corresponding decrease somewhere?

Reply:
Yes. `maybeFinish`

Reply:
I don't see it as part of the PR diff, so I am assuming you mean it was already there somehow and it's correctly counting down.